feat: version 14

murilo09 opened the pull request:
## Description

Support to version 14+

- [ ] Missing UI/Module from 14.12 charms window

## Type of change

  - [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

  - [ ] My code follows the style guidelines of this project
  - [ ] I have performed a self-review of my own code
  - [ ] I checked the PR checks reports
  - [ ] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [ ] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works
  
  Co-Authored-By: Eduardo Dantas [eduardo.dantas@hotmail.com.br](mailto:eduardo.dantas@hotmail.com.br)

dudantas committed:
init: 14.05 support

murilo09 committed:
fix: protocol 14

murilo09 committed:
fix login 14

murilo09 committed:
send appearances hash

dudantas committed:
improve: add support to fmt format to logger class

Can use:
std::string message = "Testing fmt::format";
g_logger.info("{}", message)

dudantas committed:
fix: vs warnings

dudantas committed:
fix: final 14.12 support

murilo09 committed:
Merge branch 'main' into dudantas/version-1405

kokekanon commented on src/framework/net/inputmessage.h:
```diff
@@ -32,7 +32,7 @@ class InputMessage final : public LuaObject
     {
         BUFFER_MAXSIZE = 65536,
-        MAX_HEADER_SIZE = 8
+        MAX_HEADER_SIZE = 7
```
this will break other versions below 14

kokekanon commented on src/framework/net/outputmessage.h:
```diff
@@ -33,7 +33,7 @@ class OutputMessage final : public LuaObject
         BUFFER_MAXSIZE = 65536,
         MAX_STRING_LENGTH = 65536,
-        MAX_HEADER_SIZE = 8
+        MAX_HEADER_SIZE = 7
```
this will break other versions below 14

dudantas replied:
Fixed here: https://github.com/mehah/otclient/commit/a3e69e5773b82f674aaa91c8604cc651ec3fabf2

Thanks!

kokekanon commented on src/framework/net/protocol.cpp:
```diff
@@ -179,7 +193,7 @@ void Protocol::internalRecvHeader(const uint8_t* buffer, const uint16_t size)
     // read message size
     m_inputMessage->fillBuffer(buffer, size);
-    const uint16_t remainingSize = m_inputMessage->readSize();
+    const uint16_t remainingSize = m_inputMessage->readSize() * 8 + 4;
```
this will break other versions below 14
maybe:
```cpp
    const uint16_t remainingSize = g_game.getClientVersion() >= 1400 
        ? m_inputMessage->readSize() * 8 + 4 
        : m_inputMessage->readSize();
```

kokekanon commented on src/framework/net/protocol.cpp:
```diff
@@ -316,7 +348,7 @@ void Protocol::xteaEncrypt(const OutputMessagePtr& outputMessage) const
 
     for (uint32_t i = 0, sum = 0, next_sum = sum + delta; i < 32; ++i, sum = next_sum, next_sum += delta) {
-        apply_rounds(outputMessage->getDataBuffer() - 2, encryptedSize, [&](uint32_t& left, uint32_t& right) {
+        apply_rounds(outputMessage->getHeaderBuffer(), encryptedSize, [&](uint32_t& left, uint32_t& right) {
```
this will break other versions below 14

maybe
```cpp
    XXX_t buffer = g_game.getClientVersion() >= 1400 
        ? outputMessage->getHeaderBuffer() 
        : outputMessage->getDataBuffer() - 2;
```

dudantas replied:
Fixed here: https://github.com/mehah/otclient/commit/2259250a5d5e411e00deda67a5d840575e40cf9e

Thanks!

kokekanon commented on init.lua:
```diff
@@ -9,20 +9,24 @@ Services = {
 
---[[
+
 Servers_init = {
```
This table `Servers_init` must be deactivated before being merged into the main repo

dudantas replied:
> This table `Servers_init` must be deactivated before being merged into the main repo

Yes, we will revert when everything is ok, this makes it easier to test.

kokekanon commented on src/framework/net/protocol.cpp:
```diff
@@ -28,6 +28,7 @@
 #include "inputmessage.h"
 #include "outputmessage.h"
 #include "framework/core/graphicalapplication.h"
+#include "client/game.h"
```
I think 
This implementation could be achieved without including `#include "client/game.h`". 

A new `enum GameFeature` should be created in `const.h`, and also defined in `gamelib/const.lua`. To activate it, it must be enabled in `game_feature` and then activated in `gamelib/protocollogin.lua`



example:
`m_xteaEncryptionEnabled` , `m_checksumEnabled` , `m_sequencedPackets` change the boolean from lua
https://github.com/mehah/otclient/blob/085a7876ba8164cb8eb08385812b12ceeec007aa/modules/gamelib/protocollogin.lua#L135-L148


in order to be able to login in protocol 11.00 in canary 14.xx

kokekanon commented:
if anyone has  `fmt error`
```xml
1> vc17\vcpkg_installed\x64-windows-static\include\fmt\base.h(458,28): error C2338: static_assert failed: 'Unicode support requires compiling with /utf-8'
1>(compiling source file 'otclient/x64/OpenGL/unity_QAHRA73EM9MJH98D.cpp')
 ```
 add  https://github.com/mehah/otclient/blob/main/vc17/otclient.vcxproj#L261-L262
```xml
 <AdditionalOptions>/utf-8 %(AdditionalOptions)</AdditionalOptions>
 ```

in vc17otclient.vcxproj

------------
me test:

| test                   	|   	| note|
|------------------------	|---	|---	|
| 8.6                    	|   ❌	| |
| 13.40                  	|  ❌	| |
| 14.05 canary main repo 	| ✅  	| |
| 14.12 canary pr        	|   	✅| + otcr pr # 1171 |
| 15.00 public server    	|   ✅	| + otcr pr  # 1172 |

dudantas committed:
fix: 14.12 skills

dudantas committed:
fix: missing bytes from 14.12

dudantas committed:
refactor: use dynamic maxHeaderSize in OutputMessage based on client version

This commit removes the hardcoded MAX_HEADER_SIZE constant from OutputMessage
and replaces it with a dynamically initialized m_maxHeaderSize based on the
client version (7 for >= 1405, otherwise 8), similar to InputMessage.

All references to MAX_HEADER_SIZE have been removed, and buffer positions
(m_writePos, m_headerPos) are now initialized based on m_maxHeaderSize.

This ensures consistency and future-proofing for different protocol versions
that require varying header sizes.

dudantas committed:
fix: buffer initialization

dudantas committed:
fix: sonar errors

dudantas commented:
> if anyone has `fmt error`
> 
> ```
> 1> vc17\vcpkg_installed\x64-windows-static\include\fmt\base.h(458,28): error C2338: static_assert failed: 'Unicode support requires compiling with /utf-8'
> 1>(compiling source file 'otclient/x64/OpenGL/unity_QAHRA73EM9MJH98D.cpp')
> ```
> 
> add https://github.com/mehah/otclient/blob/main/vc17/otclient.vcxproj#L261-L262
> 
> ```
>  <AdditionalOptions>/utf-8 %(AdditionalOptions)</AdditionalOptions>
> ```
> 
> in vc17otclient.vcxproj
> 
> me test:
> 
> test		note
> 8.6	❌	
> 13.40	❌	
> 14.05 canary main repo	✅	
> 14.12 canary pr	✅	+ otcr pr # 1171
> 15.00 public server	✅	+ otcr pr # 1172

13.14 was already broken before, from what I saw, I'll try to fix it

About the build solution, I always use cmake, even in visual studio, so I hadn't noticed that it broke the solution, I have an open pr to add a workflow in gha to build windows with solution, thus avoiding future problems like this. Thanks

dudantas committed:
fix: version check

dudantas committed:
Merge branch 'main' into dudantas/version-1405

dudantas committed:
fix: conflicts

dudantas committed:
fix: remove windows xp definition

dudantas committed:
fix: new charm window

dudantas committed:
Merge branch 'main' into dudantas/version-1405

dudantas committed:
fix: wrong format

javiertringol reviewed (approved):
What's missing for this to be approved? It's been 37 days since Canary approved 14.05
and 2 weeks since 14.12

What's missing? How can the community help?

dudantas committed:
Merge branch 'main' into dudantas/version-1405

vllsystems reviewed (approved):
I'm using this since 13/06 and no crashs, no kicks, no breaks.

Sounds good to me.

HopeItBuilds commented:
the `Servers_init` table must be deactivated

bateunatrave reviewed (approved):
tested for a long time without problems.

dudantas commented:
> What's missing for this to be approved? It's been 37 days since Canary approved 14.05 and 2 weeks since 14.12
> 
> What's missing? How can the community help?

The 14.12 charms window (which has been completely changed) needs to be tested and fixed.

From the comments here, it's clear that it wasn't properly tested. If it had been tested, you would have seen that the charms window doesn't work. If anyone has UI/module knowledge, they can do this so we can merge it.

javiertringol commented:
I'm happy just being able to login.

Also, many times we have to use the cipsoft client — like with the Wheel or Forge — and then go back to the OTC client. I guess it'll be the same with Charms.

kokekanon commented:
I've had it done for a while in PR #1171. I made sure not to break backward compatibility.

i create two UI  (.otui)

<img width="249" height="111" alt="image" src="https://github.com/user-attachments/assets/77b71176-3311-4dca-84ca-72fb2e70d3c1" />

`< 1340`
<img width="698" height="562" alt="image" src="https://github.com/user-attachments/assets/3e4367c7-e86c-4eac-af43-a21646db035e" />
`> 14.10`

<img width="698" height="562" alt="image" src="https://github.com/user-attachments/assets/13e6a10e-f914-4f32-a3b4-a1d5ebcb70c7" />

At that time, I was just missing information on where to get the charm data from version 14.00, since it's not retrieved from the proto or server.
I'll try to finish it today.

FelipePaluco commented:
@kokekanon if you want, here is a example of the latest image from Cipsoft.

And yes, those informations like description do not come from proto or server (they're hardcoded on client-side, like the magical archive is). If you need more support regarding the image extraction, tell me. I can help you.

<img width="702" height="541" alt="image" src="https://github.com/user-attachments/assets/8383a932-f654-4cc9-b6fa-5c994c4c71d9" />

kokekanon commented:
> @kokekanon if you want, here is a example of the latest image from Cipsoft.

i don't use canary, quick test works

<img width="885" height="586" alt="image" src="https://github.com/user-attachments/assets/296e3aae-1420-4012-b012-e47e0e681f1a" />

I have to check if I broke something in 13.40
<img width="699" height="567" alt="image" src="https://github.com/user-attachments/assets/d1a8c1ec-6c40-4252-a715-550efd163f3b" />


------------

I don't like modifying PRs that aren't mine, especially if I don't have the author's permission.

I'll reopen my PR, take the 14.00 header fix, and merge it with the features and packets I already have there.
That way, it's more likely to get approved quickly for compatibility with the main Canary repo (14.12)

FelipePaluco commented:
@kokekanon looks great! If you want, commit in the same branch. Then we can adjust anything if necessary.

lmiguelm24 commented:
> Si alguien tiene `fmt error`
> 
> ```
> 1> vc17\vcpkg_installed\x64-windows-static\include\fmt\base.h(458,28): error C2338: static_assert failed: 'Unicode support requires compiling with /utf-8'
> 1>(compiling source file 'otclient/x64/OpenGL/unity_QAHRA73EM9MJH98D.cpp')
> ```
> 
> agregue https://github.com/mehah/otclient/blob/main/vc17/otclient.vcxproj#L261-L262
> 
> ```
>  <AdditionalOptions>/utf-8 %(AdditionalOptions)</AdditionalOptions>
> ```
> 
> en vc17otclient.vcxproj
> 
> mi prueba:
> 
> prueba		nota
> 8.6	❌	
> 13.40	❌	
> Repositorio principal de Canary 14.05	✅	
> 14.12 pr canario	✅	+ otro pr # 1171
> 15.00 servidor público	✅	+ otro pr # 1172


With these exchange rates does it work out at 15.00?

luanluciano93 commented:
this PR is being closed for PR merge #1171 1171
